demos/array-shell-sort/mpc.js

```diff
@@ -0,0 +1,202 @@
+        mate[i] = i;
+      }
+      permuteRandom(mate);
+      mates.push(mate);
+    }
+    return mates;
+  }
+
+  // Generates all permutations needed for sorting process
+  function permutations(offsets) {
+    var matesByOffset = [];
+    for (var i = 0; i<offsets.length; i++) {
+      var offset = offsets[i];
+      var mates = compareRegionsPreprocess(offset);
+      matesByOffset.push(mates);
+    }
+    return matesByOffset;
+  }
+
+  // Generates all offsets needed for sorting process
+  function generateOffsets(n) {
+    var offsets = [];
+    for (var offset = Math.floor(n/2); offset > 0; offset = Math.floor(offset / 2)) {
+      offsets.push(offset);
+    }
+    return offsets;
+  }
+
+  // compare exchange two regions of length offset each
+  function compareRegions(a, s, t, offset, mates) {
+    for (var count = 0; count < C; count++) { // Do C amount of compare-exchanges
+      for (var i = 0; i < offset; i++) {
+        var n = a.length;
+        var index1 = s+i;
+        var index2 = t+mates[count][i];
+        if (index1 < n && index2 < n) {
+          compareExchange(a, index1, index2);
+        }
+      }
+    }
+  }
+
+  //Main sorting function
+  function randomizedShellSort(a, offsets, permutationsByOffset) {
+    var n = a.length;
+    for (var j = 0; j < offsets.length; j++) {
+      var offset = offsets[j];
+      var mates = permutationsByOffset[j];
+      // do a shaker pass
+      for (var i = 0; i < n - offset; i += offset) { // compare-exchange up
+        compareRegions(a, i, i + offset, offset, mates);
+      }
+      for (i = n - offset; i >= offset; i -= offset) { // compare-exchange down
+        compareRegions(a, i - offset, i, offset, mates);
+      }
+      // do extended brick pass
+      for (i = 0; i < n - 3 * offset; i += offset) { // compare 3 hops up
+        compareRegions(a, i, i + 3 * offset, offset, mates);
+      }
+      for (i = 0; i < n - 2 * offset; i += offset) { // compare 2 hops up
+        compareRegions(a, i, i + 2 * offset, offset, mates);
+      }
+      for (i = 0; i < n; i += 2 * offset) { // compare odd-even regions
+        compareRegions(a, i, i + offset, offset, mates);
+      }
+      for (i = offset; i < n - offset; i += 2 * offset) { // compare even-odd regions
+        compareRegions(a, i, i + offset, offset, mates);
+      }
+    }
+  }
+
+  /*********************************** Main Computation ********************************************/
+  exports.compute = function (input, jiff_instance) {
+    if (jiff_instance == null) {
+      jiff_instance = saved_instance;
+    }
+    var deferred = $.Deferred();
+    var promise = deferred.promise();
+
+    var final_deferred = $.Deferred();
+    var final_promise = final_deferred.promise();
+
+    // Set-up stage (No MPC here)
+    var offsets = [];
+    var permutationsByOffset = [];
+
+    // Party 1 sends message with offsets and permutation values to all parties.
+    if (jiff_instance.id===1) {
```

Comment:
This is cool, I remember telling gati that we could do this as a way to get cheap randomness for the algorithm. I do not believe that by controlling the generation of the randomness, party 1 can learn information it is not supposed to, right? (That is not relevant in all cases because we are in the semi-honest model)

Reply:
The randomness just affects the order of the compare-exchange operations and is oblivious to the underlying data, and all parties have the same amount of knowledge due to it, so I think that it does not give any information to party 1 that they should not have. Party 1 could try to sabotage by sending different randomness to different parties and do some kind of attack based on that maybe, but in the semi-honest setting I don't see it being an issue.